colflow: fix premature stats collection and improve tracking

yuzefovich opened the pull request:
This commit achieves two goals:

1. previously, we could call `finishVectorizedStatsCollectors` prematurely
in some edge cases (e.g. when the right side of the hash join is empty,
we could collect the stats from the left input tree too early). This was
caused by the fact that we accumulate the vectorized stats collectors
into a queue which is handled by the outboxes or the root materializer.
This commit begins sharing the responsibility of the collecting stats
with the hash router too. Only the hash router needs updating to achieve
the fix because it is currently the only component that splits an input
stream. The change is also sound because when the hash router exits, all
stats collectors in its input tree should have the final info, and the
hash router can collect correct stats. Those stats are added as another
metadata object to be returned by the last to exit router output.

2. `vectorizedStatsCollectorsQueue` on the flow creator is removed in
favor of more precise tracking of which stats collectors belong to which
tree. The motivation behind this changes is that the follow-up commits
will unify the retrieval of stats and draining of the metadata sources so
that the former always occurred right before the latter.

Also this commits renames `metadataSourcesQueue` to `metadataSources` and
changes the order of arguments in a few functions.

Fixes: #56928.

Release note (bug fix): Previously, CockroachDB when collecting
execution statistics (e.g. when running EXPLAIN ANALYZE) could collect
them prematurely which would result in incorrect stats. This is now
fixed.

yuzefovich committed:
colflow: fix premature stats collection and improve tracking

This commit achieves two goals:

1. previously, we could call `finishVectorizedStatsCollectors` prematurely
in some edge cases (e.g. when the right side of the hash join is empty,
we could collect the stats from the left input tree too early). This was
caused by the fact that we accumulate the vectorized stats collectors
into a queue which is handled by the outboxes or the root materializer.
This commit begins sharing the responsibility of the collecting stats
with the hash router too. Only the hash router needs updating to achieve
the fix because it is currently the only component that splits an input
stream. The change is also sound because when the hash router exits, all
stats collectors in its input tree should have the final info, and the
hash router can collect correct stats. Those stats are added as another
metadata object to be returned by the last to exit router output.

2. `vectorizedStatsCollectorsQueue` on the flow creator is removed in
favor of more precise tracking of which stats collectors belong to which
tree. The motivation behind this changes is that the follow-up commits
will unify the retrieval of stats and draining of the metadata sources so
that the former always occurred right before the latter.

Also this commits renames `metadataSourcesQueue` to `metadataSources` and
changes the order of arguments in a few functions.

Release note (bug fix): Previously, CockroachDB when collecting
execution statistics (e.g. when running EXPLAIN ANALYZE) could collect
them prematurely which would result in incorrect stats. This is now
fixed.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/62255)
<!-- Reviewable:end -->

yuzefovich commented:
This PR is a slightly updated version of #62179 which achieves the second goal mentioned in the commit message.

To provide more context why I decided to go down that route: I began implementing Alfonso's suggestion to introduce an invariant checker that all stats are collected before `DrainMeta` is called and ran into multiple issues due to the fact that we drain metadata sources in more places that in which we finish the stats collectors. I started putting some hacks in place and realized that hard-to-reason-about code was getting even worse, so I started thinking about refactoring the code (prototype in #62221) in which we carefully track stats collectors, metadata sources, and closers for each input tree, with all of those components have the same "lifecycle".

Essentially in order to get the stats invariants checker working, we need to have the same behavior for the stats collectors as we added in 2d4ad44d13d90f460c2aa431703cc2cf6518820b for the metadata sources (the materializers for wrapped processors acquiring the ownership), so I think it'll be beneficial to generalize that.

To that end #62221 will attempt the following:
- a new struct is introduced (it is essentially the same as `opDAGWithMetaSources` and will also replace `SynchronizerInput`):
```
type OpWithMetaInfo struct {
  Root colexecop.Operator
  StatsCollectors []colexecop.VectorizedStatsCollector
  MetadataSources execinfrapb.MetadataSources
  ToClose colexecop.Closers
}
```
It will track a root of the tree along with "meta" objects the ownership of which hasn't been claimed yet
- all components that currently drain metadata sources will also be responsible for collecting stats, right before draining
- a stats invariants checker is introduced which asserts that the stats collectors are finished before `DrainMeta`. This will mean that synchronizers will also be responsible for collecting stats in the input goroutines.

I think such design will lead to much cleaner and more maintainable code overall. However, the changes are pretty invasive, so  all of them wouldn't be backportable, so I extracted the first step together with the fix into this PR. Please let me know your thoughts.

asubiotto reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 25 of 25 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62255#-:-MWPAh4p9PfLNXoHRZEg:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/colflow/vectorized_flow.go, line 1023 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62255#-MWPAcxK1ZhXjBc3OqWy:-MWPAcxK1ZhXjBc3OqWz:b-kuourz) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3713f856fd743ba6c01eba6b5811b8941099eb4/pkg/sql/colflow/vectorized_flow.go#L1023)):*
> ```Go
> 			return err
> 		}
> 		s.vectorizedStatsCollectorsQueue = s.vectorizedStatsCollectorsQueue[:0]
> ```

Can we remove this queue?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62255#-:-MWPpsjN9MultvXyuVJj:b-oduykh)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/sql/colflow/vectorized_flow.go, line 1023 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/62255#-MWPAcxK1ZhXjBc3OqWy:-MWPpapr9YeDC-S6RBn4:b-8m9dlk) ([raw file](https://github.com/cockroachdb/cockroach/blob/f3713f856fd743ba6c01eba6b5811b8941099eb4/pkg/sql/colflow/vectorized_flow.go#L1023)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Can we remove this queue?
</blockquote></details>

It is removed in this commit, yes. If your question is whether it is safe to do so, then I believe it is because we're extracting the tracking of the stats collectors to be more explicit, and not as a "queue".


<!-- Sent from Reviewable.io -->
